### PR TITLE
more tests to increase code coverage

### PR DESCRIPTION
**Pull request by staltz**
This PR does a significant rewrite and improvement of the tests, particularly to increase code coverage.

**Coverage before**

Screenshot from 2021-11-15 22-18-15@2x

**Coverage after**

Screenshot from 2021-11-15 22-19-01@2x

--------

**compare()**

**Coverage before**

Screenshot from 2021-11-15 22-18-36@2x

**Coverage after**

Screenshot from 2021-11-15 22-18-52@2x

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -337,7 +337,7 @@ function compareString(buffer, start, target) {
   if (start === -1) return null
   target = Buffer.isBuffer(target) ? target : Buffer.from(target)
   var tag = varint.decode(buffer, start)
-  if (tag & (TAG_MASK !== STRING)) return null
+  if ((tag & TAG_MASK) !== STRING) return null
```
Comment (staltz) [inline index.js:340]: This definitely seemed like a mistake, and one of the new tests caught this.

File: index.js [hunk 2]
```diff
@@ -387,7 +387,7 @@ function compare(buffer1, start1, buffer2, start2) {
       buffer1.readInt32LE(start1 + len1) - buffer2.readDoubleLE(start2 + len2)
     )
 
-  if (type1 === DOUBLE && type1 === INT)
+  if (type1 === DOUBLE && type2 === INT)
```
Comment (staltz) [inline index.js:390]: This also seemed like a mistake. Read carefully just to be sure.

File: index.js [hunk 3]
```diff
@@ -404,11 +404,6 @@ function compare(buffer1, start1, buffer2, start2) {
       buffer1.readInt32LE(start1 + len1) - buffer2.readInt32LE(start2 + len2)
     )
 
-  //except for strings, sort shorter values first
-  if (type1 !== STRING) {
-    var result = type1 - type2
-    if (result) return result
-  }
```
Comment (staltz) [inline index.js:411]: If you look at line 395, then it seems that when the control flow reaches line 408, then we for sure have `type1 === type2` so it doesn't make sense to subtract in line 409 because line 410 will never trigger.

File: index.js (hunk 1 above)
Comment (arj03) [reply index.js:340]: Yes this definitely looks incorrect!

File: index.js (hunk 2 above)
Comment (arj03) [reply index.js:390]: Yep, the return line below it makes it clear that this looks good

File: index.js (hunk 3 above)
Comment (arj03) [reply index.js:411]: right

File: test/index.js [hunk 4]
```diff
@@ -63,11 +63,9 @@ testEncodeDecode({ 1: true })
 
 tape('seekPath', (t) => {
   const path = ['dependencies', 'varint']
-  const pathBuf = Buffer.alloc(bipf.encodingLength(path))
-  bipf.encode(path, pathBuf, 0)
+  const pathBuf = bipf.allocAndEncode(path)
```
Comment (arj03) [inline test/index.js:66]: yeah I was looking at that function in one of the other PRs and thought it would be quite helpful :)

**Review by arj03 — APPROVED** [review]
Great way to improve the code with a lot of extra tests. The changes in index.js looks correct. I double checked it. Even if they are corner cases it's really good to have the code super solid. Great work!
